content/posts/2025-01-30-ebpf-flows-filtering.md

```diff
@@ -0,0 +1,229 @@
+| sampling    | sampling rate to use for filtered flows                  | for example 10 or 20 (any value >= 1)                                   |           |
+| tcpFlags    | TCP flags to filter flows by                             | "SYN";"SYN-ACK";"ACK";"FIN";"RST";"URG";"ECE";"CWR";"FIN-ACK";"RST-ACK" |           |
+
+
+Note:
+
+- A rule cannot include both ports and either sourcePorts or destPorts simultaneously.
+- The number of eBPF flow filter rules is capped at `16` to maintain efficient memory usage when this feature is enabled.
+- Validation webhook rejects rules containing duplicate CIDRs.
+- Supports both IPv4 and IPv6 formats.
+- If users wish to match any CIDR, the default `0.0.0.0/0` (Null CIDR) can be used.
+
+The example below demonstrates various filtering options using multi-rule filters.
+
+```yaml
+agent:
+  type: eBPF
+  ebpf:
+    flowFilter:
+      enable: true
+      rules:
+        - action: Accept
+          cidr: 10.128.0.0/24
+          peerCIDR: 10.129.0.0/24
+          ports: '443,6443'
+          protocol: TCP
+          sampling: 10
+        - action: Accept
+          cidr: 10.129.0.1/24
+          ports: 53
+          protocol: UDP
+          sampling: 20
+        - action: Reject
+          tcpFlags: SYN
+          cidr: 10.130.0.0/24
+          protocol: TCP
+          sourcePorts: 80-100
+        - action: Accept
+          cidr: 172.30.0.0/16
+          protocol: SCTP
+          pktDrops: true
+        - action: Reject
+          cidr: 8.8.8.8/32
+          protocol: ICMP
+          icmpType: 8 // ICMP Echo request packet
+```
+2️⃣ Packet Inspection – Extract relevant packet attributes within an eBPF program.
+
+3️⃣ Early Flow Filtering –
+Skip or Allow packets based on predefined conditions before doing further ebpf packets processing.
+The Following Diagram shows how eBPF filtering is done
+
+<img src="{page.image('ebpf-flows-filtering/ebpf-flows-filtering.png')}" alt="eBPF Flows Filtering Processing">
+
+## Key Use Cases
+
+🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.
+
+🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.
+
+🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.
+
+### Filter EastWest and NorthSouth flows
+
+Allow traffic to service IP 172.30.100.64:80 with a sampling rate of 10.
+Permit communication between pods in the 10.128.0.0/16 and 10.129.0.0/16 subnets with a sampling rate of 20.
+Reject all other traffic.
+
+```yaml
+agent:
+  type: eBPF
+  ebpf:
+    flowFilter:
+      enable: true
+      rules:
+        - action: Accept
+          cidr: 172.30.0.0/16
+          sampling: 10
+        - action: Accept
+          cidr: 10.128.0.0/16
+          peerCIDR: 10.129.0.0/16
+          sampling: 20
```

Comment:
the problem with that rule is that it wouldn't accept pod-to-pod within the same subnet, such as 10.128.0.1 to 10.128.0.2
so perhaps it's not the best example. fwiw, 10.128.0.0/15 would be 10.128.0.0 to 10.129.255.255, so that would work.

Reply:
in OCP multi nodes setup we can have also 10.130 and 10.131 I will use \14 then to widen the range in the example